posthog/hogql_queries/insights/retention_query_runner.py

```diff
@@ -0,0 +1,380 @@
+                or self.event_query_type == RetentionQueryType.TARGET_FIRST_TIME
+                else self.old_filter.returning_entity
+            )
+        )
+
+        hogql_params = {key: ast.Constant(value=value) for key, value in params.items()}
+
+        data_query_hogql = parse_expr(date_query, placeholders=hogql_params)
+        if self.event_query_type != RetentionQueryType.TARGET_FIRST_TIME:
+            filter_expressions.append(data_query_hogql)
+
+        group_by_fields = None
+        having_expr = None
+        if self.event_query_type == RetentionQueryType.TARGET_FIRST_TIME:
+            group_by_fields = [ast.Field(chain=["target"])]
+            having_expr = data_query_hogql
+        if self.event_query_type == RetentionQueryType.RETURNING:
+            group_by_fields = [ast.Field(chain=["target"]), ast.Field(chain=["event_date"])]
+
+        query = f"SELECT {','.join(_fields)} FROM events {self.EVENT_TABLE_ALIAS}"
+        result = parse_select(
+            query,
+            placeholders={
+                **hogql_params,
+                "filter_where": ast.And(exprs=filter_expressions)
+                if len(filter_expressions) > 1
+                else filter_expressions[0],
+            },
+            timings=self.timings,
+        )
+        result.where = ast.And(exprs=filter_expressions)
+        result.group_by = group_by_fields
+        result.having = having_expr
+
+        sampling_factor = self.query.samplingFactor
+        if sampling_factor is not None and isinstance(sampling_factor, float):
+            sample_expr = ast.SampleExpr(sample_value=ast.RatioExpr(left=ast.Constant(value=sampling_factor)))
+            result.select_from.sample = sample_expr
+
+        return result
+
+    def target_field(self) -> str:
+        return "e.person_id as target"
+
+    def get_timestamp_field(self) -> str:
+        start_of_interval_sql = self.get_start_of_interval_sql(
+            source=f"{self.EVENT_TABLE_ALIAS}.timestamp",
+        )
+        if self.event_query_type == RetentionQueryType.TARGET:
+            return f"DISTINCT {start_of_interval_sql} AS event_date"
+        elif self.event_query_type == RetentionQueryType.TARGET_FIRST_TIME:
+            return f"min({start_of_interval_sql}) as event_date"
+        else:
+            return f"{start_of_interval_sql} AS event_date"
+
+    def _get_entity_query(self, entity: Entity):
+        if entity.type == TREND_FILTER_TYPE_ACTIONS and entity.id is not None:
+            action = Action.objects.get(pk=entity.id)
+            return action_to_expr(
+                action,
+            )
+        elif entity.type == TREND_FILTER_TYPE_EVENTS:
+            if entity.id is None:
+                return ast.Constant(value=True)
+            else:
+                return ast.CompareOperation(
+                    op=ast.CompareOperationOp.Eq,
+                    left=ast.Field(chain=[self.EVENT_TABLE_ALIAS, "event"]),
```

Comment:
Yeah, the `EVENT_TABLE_ALIAS` isn't really needed here. You can let HogQL do this for you, or, drop the alias completely if possible.